Do not use side_effect_expr_nondett in recursive_initialization

hannes-steffenhagen-diffblue opened the pull request:
If we use side_effect_expr_nondett then dump_c will generation function calls
to functions with names like `nondet_int` without declaration.

This is a problem because it’ll lead the C frontend to complain about
undeclared identifiers and such.

To fix this, we changed goto-harness to just not use side_effect_expr_nondett
at all (better long term fix would be to make sure `dump_c` outputs
declarations too).

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

hannes-steffenhagen-diffblue committed:
Do not use side_effect_expr_nondett in recursive_initialization

If we use side_effect_expr_nondett then dump_c will generation function calls
to functions with names like `nondet_int` without declaration.

This is a problem because it’ll lead the C frontend to complain about
undeclared identifiers and such.

To fix this, we changed goto-harness to just not use side_effect_expr_nondett
at all (better long term fix would be to make sure `dump_c` outputs
declarations too).

thk123 commented on regression/goto-harness/do-not-use-nondet-for-primitives/test.desc:
```diff
@@ -0,0 +1,17 @@
+test.c
+--function test_function --harness-type call-function
+\[test_function\.assertion\.1\] line \d+ assertion test != 42
+^EXIT=10$
```
:pick: It'd be good to have a positive regex for what it now contains (e.g. a declaration or whatever) since that documents the behaviour, and makes the test less likely to spuriously pass in the future

thk123 commented on regression/goto-harness/do-not-use-nondet-for-primitives/test.desc:
```diff
@@ -0,0 +1,17 @@
+--
+nondet_signed_int
+--
+If we use side_effect_expr_nondett then dump_c will generation function calls
+to functions with names like nondet_signed_int without declaration.
+This can cause problems around usability.
+
+To fix this, we changed goto-harness to just not use side_effect_expr_nondett
+at all.
+
+This test tests for the absence the use for these things.
```
:pick: This seems like great info for a commit message (since it is talking about why we moved from one solution to another) but in a test, where the reader won't have the context that we used to use `side_effect_expr_nondett` it is maybe confusing. I'd focus instead on what the test is testing - in this case not using undeclared functions.

thk123 replied:
I note this info _is_ in the commit too :+1: I still think probably more confusing than helpful in the test, but up to you

thk123 commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -169,11 +169,11 @@ void recursive_initializationt::initialize(
           const auto proper_init_case = code_function_callt{
             fun_symbol.symbol_expr(), {depth, address_of_exprt{lhs}}};
-
+          const auto should_make_equal =
```
:no_entry_sign: Looks like this is a separate case and therefore should have a separate test? :safety_pin:

thk123 commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -693,9 +693,12 @@ code_blockt recursive_initializationt::build_pointer_constructor(
   code_blockt null_and_return{{assign_null, code_returnt{}}};
   body.add(code_ifthenelset{conjunction(should_not_recurse), null_and_return});
 
+  const auto should_recurse_nondet =
```
:no_entry_sign: :safety_pin:

thk123 commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -753,8 +756,10 @@ code_blockt recursive_initializationt::build_array_string_constructor(
   {
     index_exprt index_expr{dereference_exprt{result},
                            from_integer(index, size_type())};
-    body.add(code_assignt{
```
:no_entry_sign: :safety_pin:

thk123 commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -1029,10 +1035,6 @@ code_blockt recursive_initializationt::build_function_pointer_constructor(
   const auto function_pointer_selector =
     get_fresh_local_symexpr("function_pointer_selector");
   body.add(code_declt{function_pointer_selector});
-  body.add(
```
:safety_pin:

hannes-steffenhagen-diffblue replied:
Ehem, upon trying to come up with a test for this I actually found that there was a mess up in some earlier refactoring leading to this function never ever getting called (instead implementation of what this used to do has been moved into `build_dynamic_array_constructor`.

That is a tad unfortunate. I will add a commit that removes this function to show that it doesn’t work right now, and then we can re-refactor it later if we like.

thk123 reviewed (commented):
There are a bunch of changes - it'd be good to have a test for each one to ensure the change has worked as expected. Other than that, lgtm

NlightNFotis reviewed (approved):
Good enough for me, can we have a ticket for the better long term fix?

martin-cs reviewed (approved):
Fine modulo @thk123 's comments.

hannes-steffenhagen-diffblue committed:
Remove dead method from recursive initialization

The ::build_array_string_constructor method was originally intended to be used
for building C string constructors. Unfortunately as written the condition for
calling it was actually impossible to satisfy. This wasn’t noticed so far
because similar code is now generated in ::build_dynamic_array_constructor
instead.

Dead code is a maintenance burden so it’s removed here.

hannes-steffenhagen-diffblue commented:
Note that there’s a second commit removing a dead (as in, impossible to reach) function now.
